### DISCUSSION TITLE
How to compile files in addition to .py files into the micropython firmware?

### DISCUSSION THREAD

**Discussion by iBoundary** [micropython/micropython#9587] [General]
How to compile files in addition to .py files into the micropython firmware?For example, I need to compile the Pi16x16 .raw file into the firmware.

**Comment by puppet13th** [comment]
what port/board are you trying to compile?
in case of esp32 add .py files to micropython/ports/esp32/modules

**Reply by iBoundary** [reply]
> what port/board are you trying to compile? in case of esp32 add .py files to micropython/ports/esp32/modules

I need to compile the image file into the firmware, not the python file.

**Reply by puppet13th** [reply]
do you mean to add data(non .py files like image,txt,etc) into  the firmware?

**Reply by iBoundary** [reply]
> do you mean to add data(non .py files like image,txt,etc) into the firmware?
Yes, have you tried?

**Reply by puppet13th** [reply]
i have not tried to add non .py files to the firmware. will try later and report back.

**Reply by jimmo** [reply]
> in case of esp32 add .py files to micropython/ports/esp32/modules

@puppet13th The modules directory is an implementation detail of the MicroPython build. Rather than copying your own files there you should be writing a manifest file (and even better, a custom board definition).

Please see https://docs.micropython.org/en/latest/reference/manifest.html

@iBoundary At the moment, you can only freeze .py files. I plan to add a way to add "resources" to the manifest file (with the option to make them appear on the filesystem or as a const bytes object). In the meantime, please see https://github.com/peterhinch/micropython_data_to_py

**Reply by puppet13th** [reply]
> > in case of esp32 add .py files to micropython/ports/esp32/modules
> 
> @puppet13th The modules directory is an implementation detail of the MicroPython build. Rather than copying your own files there you should be writing a manifest file (and even better, a custom board definition).
> 
> Please see https://docs.micropython.org/en/latest/reference/manifest.html

in my case, i added modules like aioble to my build. it's loads alot faster from firmware compared to vfs mpy file.

**Reply by iBoundary** [reply]
> > in case of esp32 add .py files to micropython/ports/esp32/modules
> 
> @puppet13th The modules directory is an implementation detail of the MicroPython build. Rather than copying your own files there you should be writing a manifest file (and even better, a custom board definition).
> 
> Please see https://docs.micropython.org/en/latest/reference/manifest.html
> 
> @iBoundary At the moment, you can only freeze .py files. I plan to add a way to add "resources" to the manifest file (with the option to make them appear on the filesystem or as a const bytes object). In the meantime, please see https://github.com/peterhinch/micropython_data_to_py

Okay, thank you very much.

**Reply by jimmo** [reply]
> in my case, i added modules like aioble to my build. it's loads alot faster from firmware compared to vfs mpy file.

Yes, but you should do this with a manifest.py rather than copying aioble to the modules directory. You can just `require("aioble")` (or `require("aioble-peripheral")` rather than copying files around.

**Reply by puppet13th** [reply]
> > in my case, i added modules like aioble to my build. it's loads alot faster from firmware compared to vfs mpy file.
> 
> Yes, but you should do this with a manifest.py rather than copying aioble to the modules directory. You can just `require("aioble")` (or `require("aioble-peripheral")` rather than copying files around.

thank you for your explanation, i will try to implement manifest in my build.
